// src_day_3/Form.js
import { useState } from 'react';

function Form() {
  const [message, setMessage] = useState({
    name: '',
    subject: '',
    body: '',
    reason: 'feedback',
    bestContact: '',
    acceptedTerms: false,
  });

  const handleChange = evt => {
    // Check control type.
    const value =
      evt.target.type === 'checkbox' ? evt.target.checked : evt.target.value;

    const clone = { ...message };
    clone[evt.target.name] = value;
    setMessage(clone);

    // Debugging...
    console.log(clone);
  };

  return (
    <div>
      <form>
        <div className="mb-3">
          <label htmlFor="name">Name</label>
          <input
            type="text"
            id="name"
            name="name"
            className="form-control"
            value={message.name}
            onChange={handleChange}
          />
        </div>
        <div className="mb-3">
          <label htmlFor="subject">Subject</label>
          <input
            type="text"
            id="subject"
            name="subject"
            className="form-control"
            value={message.subject}
            onChange={handleChange}
          />
        </div>
        <div className="mb-3">
          <label htmlFor="body">Message</label>
          <textarea
            id="body"
            name="body"
            className="form-control"
            value={message.body}
            onChange={handleChange}
          ></textarea>
          <label htmlFor="reason">Reason</label>
          <select
            id="reason"
            name="reason"
            className="form-control"
            value={message.reason}
            onChange={handleChange}
          >
            <option value="service">Request Service</option>
            <option value="issue">Report an Issue</option>
            <option value="sales">Sales Inquiry</option>
            <option value="feedback">General Feedback</option>
          </select>
          <div className="col">
            <input
              type="radio"
              value="morning"
              id="morning"
              name="bestContact"
              checked={message.bestContact === 'morning'}
              onChange={handleChange}
            />
            <label htmlFor="morning"> Morning</label>
          </div>
          <div className="col">
            <input
              type="radio"
              value="afternoon"
              id="afternoon"
              name="bestContact"
              checked={message.bestContact === 'afternoon'}
              onChange={handleChange}
            />
            <label htmlFor="afternoon"> Afternoon</label>
          </div>
          <div className="col">
            <input
              type="radio"
              value="evening"
              id="evening"
              name="bestContact"
              checked={message.bestContact === 'evening'}
              onChange={handleChange}
            />
            <label htmlFor="evening"> Evening</label>
          </div>
        </div>
        <div className="form-check">
          <input
            className="form-check-input"
            type="checkbox"
            id="acceptedTerms"
            name="acceptedTerms"
            checked={message.acceptedTerms}
            onChange={handleChange}
          />
          <label className="form-check-label" htmlFor="acceptedTerms">
            I've read and accept the terms and conditions.
          </label>
        </div>
      </form>
      <div>
        Your name is {message.name}. The subject is {message.subject}. Message:{' '}
        {message.body}. Feedback: {message.reason}
      </div>
      <p>Best Contact: {message.bestContact}</p>
      <p>Accepted Termas: {message.acceptedTerms}</p>
    </div>
  );
}

export default Form;
