// src_capstone_jwt_solution/App.js
import { useState } from 'react';
import Container from 'react-bootstrap/Container';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import Consoles from './Consoles';
import Finalize from './FinalizeOrder';
import Games from './Games';
import Login from './Login';
import Navigation from './Navigation';
import service from './service';
import Shirts from './Shirts';

function App() {
  const [role, setRole] = useState(service.role);
  const [order, setOrder] = useState([]);

  function handleQtyChange(event) {
    setOrder(prev => {
      const updatedItem = {
        ...prev.find(
          ({ orderId }) =>
            orderId === event.target.closest('tr').dataset.orderId
        ),
        ...{ numOrdered: event.target.value },
      };

      return [
        ...prev.filter(
          ({ orderId }) =>
            orderId !== event.target.closest('tr').dataset.orderId
        ),
        updatedItem,
      ];
    });
  }

  function handleLogin() {
    setRole(service.role);
  }

  function handleOrder(item) {
    // Change `numOrdered` or add item?
    setOrder(prev => {
      const item2UpdateQty = prev.find(
        ({ orderId }) => orderId === item.orderId
      );

      if (item2UpdateQty) {
        const updatedItem = {
          ...item2UpdateQty,
          ...{ numOrdered: item2UpdateQty.numOrdered + item.numOrdered },
        };

        const otherItems = prev.filter(
          ({ orderId }) => orderId !== item.orderId
        );
        return [...otherItems, updatedItem];
      }

      return [...prev, item];
    });
  }

  function finalizeOrder() {
    setOrder(() => []);
  }

  function logout() {
    service.logout();
    setRole('');
    setOrder(() => []);
  }

  return (
    <Router>
      <Navigation role={role} logoutHandler={logout} items={order} />
      <Container>
        <Switch>
          <Route path="/" exact>
            {role === '' ? (
              <Login handler={handleLogin} />
            ) : (
              <Games role={role} handler={handleOrder} />
            )}
          </Route>
          <Route path="/shirts" exact>
            {role === '' ? (
              <Login handler={handleLogin} />
            ) : (
              <Shirts role={role} handler={handleOrder} />
            )}
          </Route>
          <Route path="/consoles" exact>
            {role === '' ? (
              <Login handler={handleLogin} />
            ) : (
              <Consoles role={role} handler={handleOrder} />
            )}
          </Route>
          <Route path="/finalize" exact>
            {role === '' ? (
              <Login handler={handleLogin} />
            ) : (
              <Finalize
                items={order}
                handler={finalizeOrder}
                qtyHandler={handleQtyChange}
              />
            )}
          </Route>
        </Switch>
      </Container>
    </Router>
  );
}

export default App;
